src/mt.rs

```diff
@@ -67,20 +75,84 @@ impl Location {
     }
 }
 
+/// Configure a meta-tracer.
+pub struct MTBuilder {
+    hot_threshold: HotThreshold,
+}
+
+impl MTBuilder {
+    /// Create a meta-tracer with default parameters.
+    pub fn new() -> Self {
+        MTBuilder {
+            hot_threshold: DEFAULT_HOT_THRESHOLD,
+        }
+    }
+
+    /// Consume the `MTBuilder` and create a meta-tracer, returning the
+    /// [`MTThread`](struct.MTThread.html) representing the current thread.
+    pub fn init(self) -> MTThread {
+        MT::init(self.hot_threshold)
+    }
+
+    /// Change this meta-tracer builder's `hot_threshold` value.
+    pub fn hot_threshold(mut self, hot_threshold: HotThreshold) -> Self {
+        self.hot_threshold = hot_threshold;
+        self
+    }
+}
+
 /// A meta-tracer.
 pub struct MT {
     hot_threshold: HotThreshold,
+    active_threads: usize,
 }
 
 impl MT {
-    /// Create a new `MT` with default settings.
-    pub fn new() -> Self {
-        Self::new_with_hot_threshold(DEFAULT_HOT_THRESHOLD)
+    /// Create a new `MT`.
+    fn init(hot_threshold: HotThreshold) -> MTThread {
+        let mt = Arc::new(Mutex::new(Self {
+            hot_threshold,
+            active_threads: 1,
+        }));
+        MTThread::new(mt)
     }
 
-    /// Create a new `MT` with a specific hot threshold.
-    pub fn new_with_hot_threshold(hot_threshold: HotThreshold) -> Self {
-        Self { hot_threshold }
+    /// Create a new thread that can be used in the meta-tracer: the new thread that is created is
+    /// handed a [`MTThread`](struct.MTThread.html) from which the `MT` itself can be accessed.
+    pub fn spawn<F, T>(mt: Arc<Mutex<MT>>, f: F) -> io::Result<JoinHandle<T>>
+    where
+        F: FnOnce(MTThread) -> T,
+        F: Send + UnwindSafe + 'static,
+        T: Send + 'static,
+    {
+        let mt_cl = mt.clone();
+        thread::Builder::new().spawn(move || {
+            mt.lock().unwrap().active_threads += 1;
+            let r = catch_unwind(|| f(MTThread::new(mt_cl)));
+            mt.lock().unwrap().active_threads -= 1;
+            match r {
+                Ok(r) => r,
+                Err(e) => resume_unwind(e),
+            }
+        })
+    }
+}
+
+/// A meta-tracer aware thread.
+pub struct MTThread {
+    mt: Arc<Mutex<MT>>,
+    hot_threshold: HotThreshold,
+}
+
+impl MTThread {
+    fn new(mt: Arc<Mutex<MT>>) -> Self {
+        let hot_threshold = mt.lock().unwrap().hot_threshold;
```

Comment:
So there is a concious decision here to copy the threshold so that we don't have to keep unlocking the mutex to check if we exceeded the threshold? Am I right?

Reply:
Correct.